testflinger_agent/__init__.py

```diff
@@ -26,6 +31,87 @@
 logger = logging.getLogger(__name__)
 
 
+class ReqBufferTimer(Timer):
+    """Requests buffer flush"""
+
+    def run(self):
+        """Loop timer"""
+        while not self.finished.wait(self.interval):
+            self.function(*self.args, **self.kwargs)
+
+
+class ReqBufferHandler(logging.Handler):
+    """Requests logging handler"""
+
+    def __init__(self, agent, server):
+        super().__init__()
+        uri = urljoin(server, "/v1/agents/data/")
+        self.url = urljoin(uri, agent)
+        self.qdepth = 100  # messages
+        self.buffer = deque([], maxlen=self.qdepth)
+        self.reqbuf_timer = None
+        self.reqbuf_interval = 10.0  # seconds
+        self._start_rb_timer()
```

Comment:
Looks like this still runs continuously in the background rather than just calling flush on a delay when needed?

Reply:
It is instantiated, but the thread is not running (inactive) in the time between the last function/method execution and through the interval period until the next cycle.
It is blocked via threading.Event.wait() (event loop) and is inactive/blocked until the timer expires.

I suppose you could move the timer logic externally to execute after the last call of flush() by manually calling the thread's run() method, but the thread will still be instantiated via the handler init().

I don't see much of a difference, as the event loop is controlled from the main thread which instantiated the timer thread (as per source in cpython/lib/threading.py).

The thread timer's lifecycle is fully contained within the handler, and its daemon flag is set to true so its bound to the instantiating thread (main), if for some reason the handler.close() method isn't successfully called on exit.